client/packages/common/src/utils/functions/ApiUtils.tsx

```diff
@@ -0,0 +1,10 @@
+export const setNullableInput = <NODE, KEY extends keyof NODE>(
+  key: KEY,
+  entity?: NODE | null | undefined
+) => {
+  if (!entity) return undefined;
+
+  if (entity == null) return { value: undefined };
+
+  return { value: entity[key] };
```

Comment:
I'm struggling to wrap my head around how it knows this type without the explicit definition having been removed:

![Uploading EB42A0A9-41B4-416F-9C02-C95A837E7839_4_5005_c.jpeg…]()

Reply:
You mean the return type? TypeScript has structural typing which means if it looks like a certain type, then it takes it as this type, e.g.

```ts
interface Test {
   value: string
}
```
is compatible with any object that has the same structure, e.g.:

```ts
const test: Test = { value: "string", someMoreValuePossibleThough: true }
```

In setNullableInput it can derive the type of `value` from `entity[key]` (or is this part unclear?)...

Reply:
Not sure about the mechanism, but I would trust Clemens explanation. But if you want to 'see' what the end result is:

![Screenshot 2023-10-19 at 4 34 18 PM](https://github.com/openmsupply/open-msupply/assets/26194949/99d32d28-b08f-4981-9a38-ff624fbaddd0)

You can see above that typescript expects a return to be of key type: value: NonNullable<NODE>[KEY]

![Screenshot 2023-10-19 at 4 36 16 PM](https://github.com/openmsupply/open-msupply/assets/26194949/d27eb279-b98f-4844-998c-99557e3bb79e)

And here is an example of error (to show that it is indeed type safe)